### DISCUSSION TITLE
getting access to q05_nUMI_factors

### DISCUSSION THREAD

**Discussion by Nasrine26** [BayraktarLab/cell2location#117] [Q&A]
Hi, 

I really like your tool! I'm following your `Mapping human lymph node cell types to 10X Visium` tutorial and wanted to know how I could access `q05_nUMI_factors` (estimated mRNA abundance from each cell type)?  When I look at my visium data, it only has 

`obsm: 'MT', 'means_cell_abundance_w_sf', 'q05_cell_abundance_w_sf', 'q95_cell_abundance_w_sf', 'spatial', 'stds_cell_abundance_w_sf'`

Does `q05_nUMI_factors` correspond to this: `adata_vis.uns['mod']['post_sample_q05']['u_sf_mRNA_factors']` ?

Thank you for your help!

The reason I'm asking this is that I'm trying to cluster based on the cell type abundance estimated with your tool. I was wondering which is better for clustering `q05_nUMI_factors` or `q05_cell_abundance_w_sf`?

**Comment by vitkl** [comment]
Hi @Nasrine26 

Thanks for using cell2location! Yes, the tutorial exports only the most used variables to `adata.obsm`, estimates of all other variables are stored in `adata_vis.uns['mod']` (as you pointed out). 

You can add `u_sf_mRNA_factors` to `obsm` using:
```python
adata.obsm["q05_mRNA_abundance_u_sf"] = mod.sample2df_obs(
                mod.samples,
                site_name="u_sf_mRNA_factors",
                summary_name='q05',
                name_prefix="mRNA_abundance",
)
```

`q05_nUMI_factors` name was used in the previous version (pymc3).

mRNA abundance is scaled by the total RNA content of every cell type, computed using the reference cell-type signatures provided to the model scaled by the difference between technologies (e.g. neurons expressing 5k UMI vs astrocytes 2k UMI, scaled by `m_g`). So these numbers are very similar to cell abundance. Looking at mRNA abundance can be useful for QC purposes: seeing that a given cell type contributes at most 20 mRNA indicates that it is unlikely to be present in the tissue (whereas cell abundance can be larger).

**Reply by Nasrine26** [reply]
Thanks @vitkl  for the prompt response. How do I obtain mod.samples (because I don't have it)? I use mod.export_posterior(..) to export the cell type abundance estimates. Then I saw in my adata.uns a field called ['post_sample_q05']['u_sf_mRNA_factors']. Does this field correspond to exporting posterior distribution summary for mRNA abundance? 

I think I'd get the samples like this?
```
samples_w_sf = mod.sample_posterior(num_samples=1000, use_gpu=True, return_samples=True,
                                    batch_size=mod.adata.n_obs,
                                    return_sites=['w_sf', 'm_g', 'u_sf_mRNA_factors'])
```

**Reply by vitkl** [reply]
> Then I saw in my adata.uns a field called ['post_sample_q05']['u_sf_mRNA_factors']. Does this field correspond to exporting posterior distribution summary for mRNA abundance?

Yes exactly!

You can also get samples as you show below (all 1000 values).

**Reply by mosquitoCat** [reply]
Hi vitkl,

Would it be possible to have a list of key/variable name changes between the current and the previous version, please? It would be really helpful. Many thanks!

**Comment by Nasrine26** [comment]
Many thanks!
